File: starter_kits/python/bot.py
#!/usr/bin/env python3
"""BlockMarket JSONL starter Bot: one-sided quoting from the public signal."""

import json
import sys
from decimal import Decimal


def decide(observation):
    reference = Decimal(observation["reference_price"])
    lower, upper = map(Decimal, observation["limits"]["price_range"])
    bid_price = max(lower, min(reference - Decimal("0.45"), upper))
    ask_price = max(lower, min(reference + Decimal("0.45"), upper))
    bid = {
        "price": format(bid_price, ".2f"),
        "quantity": "7",
    }
    ask = {
        "price": format(ask_price, ".2f"),
        "quantity": "7",
    }
    if observation["market_signal"] == "UP":
        ask = None
    else:
        bid = None
    return {
        "decision_seq": observation["block_seq"],
        "bid": bid,
        "ask": ask,
    }


def main():
    initialized = False
    for line in sys.stdin:
        message = json.loads(line)
        if message.get("type") == "init":
            if message.get("protocol") != "blockmarket-jsonl-v1":
                raise ValueError("unsupported_protocol")
            initialized = True
        elif message.get("type") == "decision":
            if not initialized:
                raise ValueError("missing_init")
            print(
                json.dumps(decide(message["observation"]), separators=(",", ":")),
                flush=True,
            )
        elif message.get("type") == "end":
            return


if __name__ == "__main__":
    main()

File: starter_kits/typescript/bot.mjs
// Directly runnable Node.js reference for the TypeScript starter Bot.
import readline from "node:readline";

function decide(observation) {
  const toCents = (value) => Math.round(Number(value) * 100);
  const fromCents = (value) => (value / 100).toFixed(2);
  const reference = toCents(observation.reference_price);
  const [lower, upper] = observation.limits.price_range.map(toCents);
  const clamp = (value) => Math.max(lower, Math.min(value, upper));
  let bid = { price: fromCents(clamp(reference - 45)), quantity: "7" };
  let ask = { price: fromCents(clamp(reference + 45)), quantity: "7" };
  if (observation.market_signal === "UP") ask = null;
  else bid = null;
  return { decision_seq: observation.block_seq, bid, ask };
}

let initialized = false;
const lines = readline.createInterface({ input: process.stdin, crlfDelay: Infinity });
for await (const line of lines) {
  const message = JSON.parse(line);
  if (message.type === "init") {
    if (message.protocol !== "blockmarket-jsonl-v1") throw new Error("unsupported_protocol");
    initialized = true;
  } else if (message.type === "decision") {
    if (!initialized) throw new Error("missing_init");
    process.stdout.write(`${JSON.stringify(decide(message.observation))}\n`);
  } else if (message.type === "end") {
    break;
  }
}
